src/java.base/share/classes/java/io/ObjectInputFilter.java

```diff
@@ -48,50 +53,163 @@
+ * }</pre>
+ *
+ * <p>In an application with multiple execution contexts, the
+ * {@linkplain Config#setSerialFilterFactory(BinaryOperator) filter factory} can better
+ * protect individual contexts by providing a custom filter for each. When the stream
+ * is constructed, the filter factory can identify the execution context based upon
+ * the current thread-local state, hierarchy of callers, library, module, and class loader.
+ * At that point, a policy for creating or selecting filters can choose a specific filter
+ * or composition of filters based on the context.
+ *
+ * <p> When a filter is set on an {@link ObjectInputStream}, the {@link #checkInput checkInput(FilterInfo)}
  * method is called to validate classes, the length of each array,
  * the number of objects being read from the stream, the depth of the graph,
  * and the total number of bytes read from the stream.
+ * The filter is invoked zero or more times
+ * while {@linkplain ObjectInputStream#readObject() reading objects}.
+ * The JVM-wide deserialization filter factory ensures that a deserialization filter can be set
+ * on every {@link ObjectInputStream} and every object read from the stream can be checked.
  * <p>
- * A filter can be set via {@link ObjectInputStream#setObjectInputFilter setObjectInputFilter}
- * for an individual ObjectInputStream.
- * A filter can be set via {@link Config#setSerialFilter(ObjectInputFilter) Config.setSerialFilter}
- * to affect every {@code ObjectInputStream} that does not otherwise set a filter.
+ * The deserialization filter for a stream is determined in one of the following ways:
+ * <ul>
+ * <li>A JVM-wide filter factory can be set via {@link Config#setSerialFilterFactory(BinaryOperator)}
+ *     or the system property {@code jdk.serialFilterFactory} or
+ *     the security property {@code jdk.serialFilterFactory}.
+ *     The filter factory is invoked for each new ObjectInputStream and
+ *     when a filter is set for a stream.
+ *     The filter factory determines the filter to be used for each stream based
+ *     on its inputs, thread context, other filters, or state that is available.
+ * <li>If a JVM-wide filter factory is not set, a builtin deserialization filter factory
+ *     provides the {@link Config#getSerialFilter static JVM-wide filter} when invoked from the
+ *     {@link ObjectInputStream#ObjectInputStream(InputStream) ObjectInputStream constructors}
+ *     and replaces the static filter when invoked from
+ *     {@link ObjectInputStream#setObjectInputFilter(ObjectInputFilter)}.
+ *     See {@link Config#getSerialFilterFactory() getSerialFilterFactory}.
+ * <li>A stream-specific filter can be set for an individual ObjectInputStream
+ *     via {@link ObjectInputStream#setObjectInputFilter setObjectInputFilter}.
+ *     Note that the filter may be used directly or combined with other filters by the
+ *     {@linkplain Config#setSerialFilterFactory(BinaryOperator) JVM-wide filter factory}.
+ * </ul>
```

Comment:
In previous versions, calling OIS.setObjectInputFilter determined exactly the filter used for the stream.
With the filter factory enhancement, the current filter factory determines how the argument to OIS.setObjectInputFilter is used. There are plenty of cases where the filter factory will combine it with other filters and the composite will becomes the filter for the stream.

Reply:
Here is the source of my confusion. The bulleted list is enumerating how a stream-specific filter is determined, but I see an extra step in that which should be unnecessary. It is currently:

1. Check JVM-wide filter factory
2. If no JVM-wide, check built-in factory
3. setSerialFilterFactory

, but 1 and 2 are not separate and distinct cases - there is always a JVM-wide deserialization filter factory. The JVM-wide deserialization filter factory is either i) set through a property, or ii) set explicitly by an API call, or iii) the built-in implementation.

If the initialisation of the JVM-wide deserialization filter factory is separated out from how the stream-specific factory is determined, then I believe that it will be easier to follow.

Reply:
Yes, I will describe the filter factory selection and separately the filter composition by the factory.